secret_provider/src/k8s.rs

```diff
@@ -0,0 +1,82 @@
+//! This module reads from a secret provided the given key matches. Consider the example as below.
+//!
+//! apiVersion: v1
+//! kind: Secret
+//! metadata:
+//!   name: my-secret
+//!   namespace: default
+//! type: Opaque
+//! stringData:
+//!   encryption_parameters: |
+//!    {
+//!       "cipher": "AesCbc",
+//!       "key": {
+//!        "key_name": "example-key",
+//!        "key": [115, 111, 109, 101, 45, 107, 101, 121],
+//!         "key_length": 128,
+//!         "key2": null,
+//!         "key2_length": null
+//!       }
+//!     }
+//!
+use crate::types::EncryptionParameters;
+use crate::SecretProvider;
+
+use crate::error::Error;
+use async_trait::async_trait;
+use k8s_openapi::api::core::v1::Secret;
+
+const SECRET_DATA_KEY: &str = "encryption_parameters";
+
+/// A secret provider that retrieves a Kubernetes secret.
+pub struct K8sSecretProvider {
+    /// The Kubernetes client.
+    pub client: kube::Client,
+    /// The namespace in which to look for the secret.
+    pub namespace: String,
+    /// The key within the secret’s data containing the encryption parameters.
+    pub secret_data_key: String,
+}
+
+impl K8sSecretProvider {
+    /// Create a new `K8sSecretProvider` using the provided Kubernetes client and namespace.
+    pub fn new(client: kube::Client, namespace: impl Into<String>) -> Self {
+        Self {
+            client,
+            namespace: namespace.into(),
+            secret_data_key: SECRET_DATA_KEY.to_string(),
+        }
+    }
+}
+
+#[async_trait]
+impl SecretProvider for K8sSecretProvider {
+    async fn get_encryption_parameters(
+        &self,
+        secret_name: &str,
+    ) -> Result<EncryptionParameters, Error> {
+        let secrets: kube::api::Api<Secret> =
+            kube::api::Api::namespaced(self.client.clone(), &self.namespace);
+
+        let secret = secrets
+            .get(secret_name)
+            .await
+            .map_err(|e| Error::KubeApi { source: e })?;
+
+        let data = secret.data.ok_or_else(|| Error::MissingSecretData {})?;
+        let raw_bytes = data
+            .get(&self.secret_data_key)
+            .ok_or_else(|| Error::MissingSecretKey {
+                key: self.secret_data_key.to_string(),
+            })?
+            .0
+            .clone();
```

Comment:
don't need to clone? The reference is used below

Reply:
Doesn't work without the clone.

Reply:
Can you clarify "doesn't work"?

Reply:
Done.